src/blockcholesky.jl

```diff
@@ -1,45 +1,83 @@
 
+
 ##########################
 # Cholesky Factorization #
 ##########################
 
 """
 Funtions to do
-1.cholesky structure
 2.check square
 3.check positive definite
-4.swap 'cholesky!'
 """
 
 cholesky(A::Symmetric{<:Real,<:BlockArray},
     ::Val{false}=Val(false); check::Bool = true) = cholesky!(cholcopy(A); check = check)
 
-function cholesky!(A::Symmetric{<:Real,<:BlockArray}; check::Bool = true)
-    
-    chol_P = parent(A)
 
-    # Initializing the first role of blocks
-    cholesky!(Symmetric(getblock(chol_P,1,1)))
-    for j = 2:blocksize(A)[1]
-        ldiv!(UpperTriangular(getblock(chol_P,1,1))', getblock(chol_P,1,j))
-    end
+function b_chol!(A::BlockArray{T}, ::Type{UpperTriangular}) where T<:Real
+    n = blocksize(A)[1]
 
-    # For the left blocks
-    for i = 2:blocksize(A)[1]
-        Pii = getblock(chol_P,i,i) 
-        for k = 1:i-1
-            mul!(Pii,getblock(chol_P,k,i)',getblock(chol_P,k,i),-1.0,1.0)
+    @inbounds begin
+        # Initializing the first role of blocks
+        cholesky!(Symmetric(getblock(A,1,1)))
+        for j = 2:n
+            ldiv!(UpperTriangular(getblock(A,1,1))', getblock(A,1,j))
         end
-        cholesky!(Symmetric(Pii))
 
-        for j = i+1:blocksize(A)[1]
-            Pij = getblock(chol_P,i,j)
+        # For the left blocks
+        for i = 2:n
+            Pii = getblock(A,i,i) 
             for k = 1:i-1
-                mul!(Pij,getblock(chol_P,k,i)',getblock(chol_P,k,j),-1.0,1.0)
+                muladd!(-1.0, getblock(A,k,i)', getblock(A,k,i), 1.0, Pii)
+            end
+            cholesky!(Symmetric(Pii))
+    
+            for j = i+1:n
+                Pij = getblock(A,i,j)
+                for k = 1:i-1
+                    muladd!(-1.0, getblock(A,k,i)', getblock(A,k,j), 1.0, Pij)
+                end
+                ldiv!(UpperTriangular(getblock(A,i,i))', Pij)
             end
-            ldiv!(UpperTriangular(getblock(chol_P,i,i))', Pij)
         end
     end
+
+    return UpperTriangular(A)

```

Comment:
Please have this return `Cholesky(A, :U, 0)`. The `0` should actually be `info` which is determined from the `cholesky!` call, so probably what we want is
```julia
info = 0
for ...
    info = cholesky!(Symmetric(Pii))
   iszero(info) || break
end
Cholesky(A, :U, info)
```
though we should double check what `info` actually means: that is, do we break and `info` indicates a problem, or do we continue as if it were SPD?

Similarly for lower below.

Reply:
In the document of `PosDefException(info)`, it says that the `info` field indicates the location of (one of) the eigenvalue(s) which is (are) less than/equal to 0. And the `info` is determined by the function `LAPACK.potrf!(uplo, A)`. So we can do the check of SPD in a similar way with `LinearAlgebra.cholesky`

Reply:
I think you mean indicates that there exists a negative eigenvalue, not that it finds one. Here is the doc for LAPACK.potrf!: http://www.netlib.org/lapack/explore-html/d1/d7a/group__double_p_ocomputational_ga2f55f604a6003d03b5cd4a0adcfb74d6.html

```
INFO is INTEGER
          = 0:  successful exit
          < 0:  if INFO = -i, the i-th argument had an illegal value
          > 0:  if INFO = i, the leading minor of order i is not
                positive definite, and the factorization could not be
                completed.
```
So I think we want 
```julia
if !iszero(info)
    @assert info > 0 # should never have illegal values
    info += sum(blockaxes(A,2)[Block.(1:i-1)]) # add columns before we got to block to indicate correct minor
    break
end
```